HW2/P5/driver.py

```diff
@@ -80,3 +97,24 @@ def randcolor():
             # SUBPROBLEM 3: sort objects by location.  Be sure to update the
             # grid if objects' indices change!  Also be sure to sort the
             # velocities with their object positions!
+            
+            # Go through the grid, find all of the non-empty spaces, 
+            # save the ids and their grid positions into a new list
+            grid_positions_list = []
+            for i in range(grid.shape[0]):
+                for j in range(grid.shape[1]):
+                    if grid[i, j] != 4294967295:
+                        grid_positions_list.append([grid[i, j], (i, j)])
+            
+            # Calculate the new order based on z_order algorithm
+            sort_order = sorted(grid_positions_list, cmp_zorder, key=lambda t: t[1])
+            
+            keys = [so[0] for so in sort_order]
+            
+            # Apply the new order to the positions and the velocities
+            positions = positions[keys]
+            velocities = velocities[keys]
+            
+            # Reset the grid
+            grid = - np.ones((grid_size, grid_size), dtype=np.uint32)
+            grid[(positions[:, 0] / grid_spacing).astype(int), (positions[:, 1] / grid_spacing).astype(int)] = np.arange(len(keys))
```

Comment:
I like the way you used numpy's indexing benefits to make this fast - I did it basically the same way.
I would recommend, as others have suggested, that you declare a variable that holds (positions[:, 0] / grid_spacing), just so it is more clear. Naturally, that has no speed benefit, but from my side, it improves readability.